src/tools/dashboard/src/litani.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Utilities to interact with the `Litani` build accumulator.
+
+use pulldown_cmark::escape::StrWrite;
+use std::collections::HashMap;
+use std::path::Path;
+use std::process::{Child, Command};
+
+/// Data structure representing a `Litani` build.
+pub struct Litani {
+    /// A buffer of the `spawn`ed Litani jobs so far. `Litani` takes some time
+    /// to execute each `add-job` command and executing thousands of them
+    /// sequentially takes a considerable amount of time. To speed up the
+    /// execution of those commands, we spawn those commands sequentially (as
+    /// normal). However, instead of `wait`ing for each process to terminate,
+    /// we add its handle to a buffer of the `spawn`ed processes and continue
+    /// with our program. Once we are done adding jobs, we wait for all of them
+    /// to terminate before we run the `run-build` command.
+    spawned_commands: Vec<Child>,
+}
+
+impl Litani {
+    /// Sets up a new [`Litani`] run.
+    pub fn init(project_name: &str, output_prefix: &Path, output_symlink: &Path) -> Self {
+        Command::new("litani")
+            .args([
+                "init",
+                "--project-name",
+                project_name,
+                "--output-prefix",
+                output_prefix.to_str().unwrap(),
+                "--output-symlink",
+                output_symlink.to_str().unwrap(),
+            ])
+            .spawn()
+            .unwrap()
+            .wait()
+            .unwrap();
+        Self { spawned_commands: Vec::new() }
+    }
+
+    /// Adds a single command with its dependencies.
+    pub fn add_job(
+        &mut self,
+        command: &Command,
+        description: &str,
+        pipeline: &str,
+        stage: &str,
+        exit_status: i32,
+    ) {
+        let mut job = Command::new("litani");
+        // The given command may contain additional env vars. Prepend those vars
+        // to the command before passing it to Litani.
+        let job_envs: HashMap<_, _> = job.get_envs().collect();
+        let mut new_envs = String::new();
+        command.get_envs().fold(&mut new_envs, |fmt, (k, v)| {
+            if !job_envs.contains_key(k) {
+                fmt.write_fmt(format_args!(
+                    "{}=\"{}\" ",
+                    k.to_str().unwrap(),
+                    v.unwrap().to_str().unwrap()
+                ))
+                .unwrap();
+            }
+            fmt
+        });
+        job.args([
+            "add-job",
```

Comment:
If you're using inputs and outputs as I suggest below, I suggest also passing in `--phony-outputs`; this is available in the develop branch of Litani, we'll release it probably next week

Reply:
That's SO COOOL!!!